### ISSUE TITLE
pynsist claims that `cryptography-3.3.1-cp36-abi-win32.whl` is not compatible with Python 3.8.6, win32

### ISSUE THREAD

**Issue by ntamas** [takluyver/pynsist#219]
I'm trying to package an app that depends on the `cryptography` library. `cryptography` provides `abi3` wheels, which are [supposed to be compatible](https://cryptography.io/en/latest/faq.html#why-are-there-no-wheels-for-my-python3-x-version) with any Python version greater than or equal to the one specified in the filename, so a `cp36-abi` wheel should be compatible with Python 3.8.

Is this a bug in `pynsist` or am I misunderstanding something about the compatibility issues?

This is the error I'm getting:

```
[...]
Collecting wheel file: cryptography-3.3.1-cp36-abi3-win32.whl (from: ./build/wheels/*.whl)
Traceback (most recent call last):
[...]
  File ".venv/lib/python3.8/site-packages/nsist/__init__.py", line 487, in run
    self.prepare_packages()
  File ".venv/lib/python3.8/site-packages/nsist/__init__.py", line 359, in prepare_packages
    wg.get_all()
  File ".venv/lib/python3.8/site-packages/nsist/wheels.py", line 299, in get_all
    self.get_globs()
  File ".venv/lib/python3.8/site-packages/nsist/wheels.py", line 317, in get_globs
    self.validate_wheel(path)
  File ".venv/lib/python3.8/site-packages/nsist/wheels.py", line 338, in validate_wheel
    raise ValueError('Wheel {} is not compatible with Python {}, {}'
ValueError: Wheel cryptography-3.3.1-cp36-abi3-win32.whl is not compatible with Python 3.8.6, win32
```

**Comment by takluyver** [comment]
The code for checking compatibility currently treats `cp33` as meaning CPython 3.3 specifically, not >= 3.3. Here's the relevant code:

https://github.com/takluyver/pynsist/blob/89431e7c382da103f08c2fe006e388db4d2cf09b/nsist/wheels.py#L47-L55

I can't currently see where in the specs it says that it should mean >=3.3. PEP 425, which defines compatibility 
tags, has an example list of compatible tags for CPython 3.3, which includes `cp33-abi3-linux_x86_64` and `cp3-abi3-linux_x86_64`, but not `cp32-abi3-linux_x86_64`.

So I'd like to see either something in a spec that clarifies that's how it's meant to be interpreted, or a discussion (e.g. on https://discuss.python.org/ ) to confirm that this how it's meant to be, not just an oddity of pip that cryptography has taken advantage of.

**Comment by ntamas** [comment]
I'm trying to wrap my head around it as well, but if you simply run `pip install cryptography` (or `bcrypt`, or many other libraries that publish `abi3` wheels) on Python 3.8 or 3.9, `pip` will happily install the `cp36-abi3` wheel, and the library can be imported without problems. I don't think that `cryptography` is explicitly whitelisted in `pip` so it must be somewhere in the rules that `cp36-abi3` is compatible with newer versions of Python. Note that other modules also use this; `bcrypt`

I think that the key is in [PEP 384](https://www.python.org/dev/peps/pep-0384/), which describes the stable ABI (i.e. `abi3`). This is in the abstract:

> This PEP proposes to define a stable set of API functions which are guaranteed to be available for the lifetime of Python 3, and which will also remain binary-compatible across versions. Extension modules and applications embedding Python can work with different feature releases as long as they restrict themselves to this stable ABI.

In this sense `abi3` as an ABI tag is different from, say, `cp36` as an ABI tag; `cp36-cp36-win32` means that the module works on `win32` with any version of Python 3.6 and above (first `cp36`) as long as it satisfies the `cp36` ABI (second `cp36`), which, incidentally, refers to Python 3.6.x _only_). `cp36-abi3-win32` means that the module works on `win32` with any version of Python 3.6 and above as long as it satisfies the `abi3` ABI, which is guaranteed to remain the same for all subsequent Python 3.x releases.

**Comment by ntamas** [comment]
@takluyver I have found [another library](https://pypi.org/project/python-json-logger/) where I think `pynsist` falsely claims that it is not compatible with a particular version.

`setup.py` in `python-json-logger` contains `python_requires='>=3.4'`, so the corresponding Python wheel (created locally on my machine with `pip wheel` using the most recent `pip` version) is named `python_json_logger-2.0.1-py34-none-any.whl`. `pynsist` says that it is "not compatible with Python 3.7.9, win32", while the original `setup.py` clearly states that 3.4 is a _minimum_ version requirement for Python, not an exact version requirement.

**Comment by ntamas** [comment]
Interestingly enough, other projects (like PyQt5) explicitly list all the Python versions in the name of the wheel; [see here for an example](https://pypi.org/project/PyQt5/#files) where the name of the wheel includes `cp35.cp36.cp37.cp38.cp39`; not sure how the current code in `pynsist2` would deal with that but it's also something worth looking into.

**Comment by takluyver** [comment]
The dot-separated list is [part of the spec](https://www.python.org/dev/peps/pep-0425/#compressed-tag-sets), and Pynsist does (I hope) handle it correctly - that's the `.split('.')` on the last line of the code I linked above.

I'll try to find time to dig into what's going on with using the `cp34` tag to mean `>= 3.4` rather than `== 3.4`, as I understand it.

**Comment by petersilva** [comment]
I get this same error message with a different package:
```
   raise ValueError('Wheel {} is not compatible with Python {}, {}' 
ValueError: Wheel PyNaCl-1.4.0-cp38-cp38-win_amd64.whl is not compatible with Python 3.8.8, win_amd64
```

I tried every version of python available 3.8.0->3.8.8 and it gives the same error for all of them.
the code is on github, so I can give people something reproducible...

Do you think it is the same problem, or should I open a new ticket?

**Comment by petersilva** [comment]
see #224

**Comment by takluyver** [comment]
I think it's ambiguous how the tags are interpreted, so I've started a discussion on the packaging section of the Python discourse forum:

https://discuss.python.org/t/python-tags-specific-version-of-interpreter-or-minimum-version/7831

**Comment by rodrigomologni** [comment]
Hi, @ntamas! The same problem happened here. I'm using cryptography and Python 3.7. I overcame this problem as follows:

1. pip download cryptography
2. rename `cryptography*cp36*.whl` to `cryptography*cp37*.whl`
3. include `local_wheels=cryptography*cp37*.whl` to `installer.cfg`

**Comment by takluyver** [comment]
I've just released 2.7, which should fix this.
